pymeasure/instruments/oxfordinstruments/ips120_10.py

```diff
@@ -0,0 +1,330 @@
+            sleep(self._SWITCH_HEATER_DELAY)
+            self.activity = "to zero"
+            self.wait_for_idle()
+        else:
+            raise Exception("Switch status returned %d" % switch_status)
+
+    def disable_persistent_mode(self):
+        """ Methods that disables the persistent magnetic field mode. """
+        # Check if system idle
+        if not self.sweep_status == "at rest":
+            raise Exception("Magnet not at rest")
+
+        # Check if the setpoint equals the persistent field
+        if not self.field == self.field_setpoint:
+            log.warning("IPS: field setpoint and persistent field not identical; "
+                        "setting the setpoint to the persistent field.")
+            self.field_setpoint = self.field
+
+        switch_status = self.switch_heater_status
+        if switch_status == 1:
+            return  # Magnet already in demand mode or at 0 field
+        elif switch_status in [0, 2]:
+            self.activity = "to setpoint"
+            self.wait_for_idle()
+            self.activity = "hold"
+            self.switch_heater_status = 1
+            log.info("IPS: Wait for 20s for switch heater")
+            sleep(self._SWITCH_HEATER_DELAY)
+        else:
+            raise Exception("Switch status returned %d" % switch_status)
+
+    def wait_for_idle(self, max_errors=10, delay=1):
+        error_ct = 0
+        status = None
+        while True:
+            sleep(delay)
+
+            try:
+                status = self.sweep_status
+            except Exception as e:
+                log.error("IPS: Issue with getting status (#%d): %s" % (error_ct, e))
+                error_ct = 0
+
+            if status == "at rest":
+                break
+
+            if error_ct > max_errors:
+                raise Exception("Too many exceptions occurred during getting IPS status.")
+
+    def set_field(self, field, sweep_rate=None, persistent_mode_control=True):
+        # Check if field needs changing
+        if self.field == field:
+            return
+
+        switch_status = self.switch_heater_status
+        if switch_status == 1:
+            pass  # Magnet in demand mode
+        elif switch_status in [0, 2]:
+            # Magnet in persistent mode
+            if persistent_mode_control:
+                self.disable_persistent_mode()
+            else:
+                raise Exception("IPS in persistent mode and control of persistent mode not allowed")
+
+        if sweep_rate is not None:
+            self.sweep_rate = sweep_rate
+
+        if field == 0:
+            self.activity = "to zero"
+        else:
+            self.activity = "to setpoint"
+            self.field_setpoint = field
+
+        self.wait_for_idle()
+        sleep(10)
+
+        if persistent_mode_control and field != 0:
+            self.enable_persistent_mode()
+
+    def train_magnet(self, training_scheme=None):
+        """ Method that trains the magnet after cooling down """
```

Comment:
Maybe define what `training_scheme` should look like in the docstring.

Reply:
I would prefer that too, over a pre-defined training scheme. The way it is now, the default scheme is good for only an (uncommonly) strong magnet, and I would much prefer it if the optional argument was given at initialisation, without an otherwise defined scheme. After all, this scheme is not a characteristic property of the device itself, but of the magnet it controls.

Reply:
You are right indeed; I've added to the docstring how the training_scheme should look like (and in the general-docstring of the class an example of how to use the instrument, including the train_magnet method. I've decided to remove the training scheme as an attribute from the class as I think you're right in stating that it is not an property of this device, and it is probably even not useful to define it for most instruments upon initialization, but only whenever used.